src/twisted/web/http.py

```diff
@@ -1825,75 +1823,97 @@ class _ChunkedTransferDecoder:
-            parts = line.split(b";")
-            try:
-                self.length = int(parts[0], 16)
-            except ValueError:
-                raise _MalformedChunkedDataError("Chunk-size must be an integer.")
-            if self.length < 0:
-                raise _MalformedChunkedDataError("Chunk-size must not be negative.")
-            elif self.length == 0:
-                self.state = "TRAILER"
-            else:
-                self.state = "BODY"
-            return rest
+    def _dataReceived_CHUNK_LENGTH(self) -> bool:
+        """
+        Read the chunk size line, ignoring any extensions.
+
+        @returns: C{True} once the line has been read and removed from
+            C{self._buffer}.  C{False} when more data is required.
+
+        @raises _MalformedChunkedDataError: when the chunk size cannot be
+            decoded.
+        """
+        eolIndex = self._buffer.find(b"\r\n", self._start)
+        if eolIndex == -1:
+            self._start = len(self._buffer) - 1
+            return False
+
+        endOfLengthIndex = self._buffer.find(b";", 0, eolIndex)
+        if endOfLengthIndex == -1:
+            endOfLengthIndex = eolIndex
+        try:
+            length = int(self._buffer[0:endOfLengthIndex], 16)
+        except ValueError:
+            raise _MalformedChunkedDataError("Chunk-size must be an integer.")
+
+        if length < 0:
+            raise _MalformedChunkedDataError("Chunk-size must not be negative.")
+        elif length == 0:
+            self.state = "TRAILER"
         else:
-            self._buffer = data
-            return b""
+            self.state = "BODY"
+
+        self.length = length
+        del self._buffer[0 : eolIndex + 2]
+        self._start = 0
+        return True
 
-    def _dataReceived_CRLF(self, data):
-        if data.startswith(b"\r\n"):
+    def _dataReceived_CRLF(self) -> bool:
+        if self._buffer.startswith(b"\r\n"):
             self.state = "CHUNK_LENGTH"
-            return data[2:]
+            del self._buffer[0:2]
+            return True
         else:
-            self._buffer = data
-            return b""
+            return False
 
-    def _dataReceived_TRAILER(self, data):
-        if data.startswith(b"\r\n"):
-            data = data[2:]
+    def _dataReceived_TRAILER(self) -> bool:
+        if self._buffer.startswith(b"\r\n"):
+            data = memoryview(self._buffer)[2:].tobytes()
+            del self._buffer[:]
             self.state = "FINISHED"
             self.finishCallback(data)
-        else:
-            self._buffer = data
-        return b""
+        return False
 
-    def _dataReceived_BODY(self, data):
-        if len(data) >= self.length:
-            chunk, data = data[: self.length], data[self.length :]
-            self.dataCallback(chunk)
+    def _dataReceived_BODY(self) -> bool:
+        if len(self._buffer) >= self.length:
+            chunk = memoryview(self._buffer)[: self.length].tobytes()
```

Comment:
Maybe a comment would help to explain what each branch is handling.

```suggestion
        if len(self._buffer) >= self.length:
            # All data received for the current chunk.
            chunk = memoryview(self._buffer)[: self.length].tobytes()
```

Reply:
I'll add a docstring, since this function lacks one.

Reply:
Thanks for the update. Just a very minor comment in terms of style.

I find it more helpful if the docstring of a function describes how,when,why to use it it and then inside comments explains what it does :)

For example thinks like `Then invoke C{finishCallback} and switch to state C{'FINISHED'}.` don't seem to useful for me as I can already read that from the code.

My issue was with the condition that was not obvious for a new developers just starting to read the code.

So maybe something like this:

```suggestion
        full_chunk_received = len(self._buffer) >= self.length
        if full_chunk_received:
            chunk = memoryview(self._buffer)[: self.length].tobytes()
```